### PR TITLE
[build] Fix wrong use of %{project_root}/_build in config files

### PR DESCRIPTION
**Pull request by ejgallego**
In particular, imagine a build setup such as:

```
my-dir/dune-project
my-dir/coq/...
my-dir/coq-lsp/...
```

dune will create `my-dir/_build`, however, inside Coq dune files,
`%{project_root}` will resolve to `my-lib/coq/_build` which doesn't
exist.

Thanks to Rodolphe Lepigre for catching this bug.

cc: #19857

### CODE REVIEW

**Comment by ejgallego** [comment]
> What is the actual bad behaviour?

Take any composed build setup such as:

```
my-dir/dune-project
my-dir/coq
my-dir/coq-lsp
```

for example, then dune will use `my-dir/_build` as directory, however, inside Coq dune files, `%{project_root}` will resolve to `my-lib/coq/_build` which doesn't exist.

I will submit a test for this in CI, I actually wrote the patch a long time ago, and would have been helpful for the work on stdlib split, but I got sadly distracted by some unplanned stuff.

**Comment by rlepigre** [comment]
Additionally, there are a couple of fixes to do in `dev/shim/dune`. Here is the diff I have:
```diff
diff --git a/dev/shim/dune b/dev/shim/dune
index 7a95da2503..116480eebc 100644
--- a/dev/shim/dune
+++ b/dev/shim/dune
@@ -18,7 +18,26 @@
   (with-stdout-to %{targets}
    (progn
     (echo "#!/usr/bin/env bash\n")
-    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqtop} -I \"$(dirname \"$0\")/%{project_root}/../install/default/lib\" -coqlib \"$(dirname \"$0\")/%{project_root}\" \"$@\"'")
+    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqtop} -I \"$(dirname \"$0\")/%{workspace_root}/_build/install/default/lib\" -coqlib \"$(dirname \"$0\")/%{project_root}\" \"$@\"'")
+    (run chmod +x %{targets})))))
+
+; coqidetop
+
+(alias
+ (name coqidetop-prelude)
+ (deps
+  %{bin:coqidetop.opt}
+   ; XXX: bug, we are missing the dep on the _install meta file...
+  %{project_root}/theories/Init/Prelude.vo))
+
+(rule
+ (targets coqidetop.opt)
+ (deps (alias coqidetop-prelude))
+ (action
+  (with-stdout-to %{targets}
+   (progn
+    (echo "#!/usr/bin/env bash\n")
+    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqidetop.opt} -I \"$(dirname \"$0\")/%{workspace_root}/_build/install/default/lib\" -coqlib \"$(dirname \"$0\")/%{project_root}\" \"$@\"'")
     (run chmod +x %{targets})))))
 
 ; coqc
@@ -37,7 +56,7 @@
   (with-stdout-to %{targets}
    (progn
     (echo "#!/usr/bin/env bash\n")
-    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqc} -I \"$(dirname \"$0\")/%{project_root}/../install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} -nI \"$(dirname \"$0\")\"/%{project_root}/kernel/.kernel.objs/byte \"$@\"'")
+    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqc} -I \"$(dirname \"$0\")/%{workspace_root}/_build/install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} -nI \"$(dirname \"$0\")\"/%{project_root}/kernel/.kernel.objs/byte \"$@\"'")
     (run chmod +x %{targets})))))
 
 ; coqtop.byte
@@ -80,7 +99,7 @@
   (with-stdout-to %{targets}
    (progn
     (echo "#!/usr/bin/env bash\n")
-    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqtop.byte} -I \"$(dirname \"$0\")/%{project_root}/../install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} \"$@\"'")
+    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqtop.byte} -I \"$(dirname \"$0\")/%{workspace_root}/_build/install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} \"$@\"'")
     (run chmod +x %{targets})))))
 
 ; coqide
@@ -91,7 +110,7 @@
   ; without this if the gtk libs are not available dune can try to use
   ; coqide from PATH instead of giving a nice error
   ; there is no problem with the other shims since they don't depend on optional build products
-  %{project_root}/ide/coqide/coqide_main.exe
+  %{project_root}/ide/coqide/rocqide_main.exe
   %{bin:coqworker.opt}
   %{project_root}/theories/Init/Prelude.vo
   %{project_root}/coqide-server.install
@@ -104,5 +123,5 @@
   (with-stdout-to %{targets}
    (progn
     (echo "#!/usr/bin/env bash\n")
-    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqide} -I \"$(dirname \"$0\")/%{project_root}/../install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} \"$@\"'")
+    (bash "echo '\"$(dirname \"$0\")\"/%{bin:coqide} -I \"$(dirname \"$0\")/%{workspace_root}/_build/install/default/lib\" -coqlib \"$(dirname \"$0\")\"/%{project_root} \"$@\"'")
     (run chmod +x %{targets})))))
```
This also adds a shim for `coqidetop.opt` as well as fixes a bad reference to `coqide_main.exe` (renamed to `rocqide_main.exe`).

**Comment by rlepigre** [comment]
I forgot to say: thanks for the MR @ejgallego!

**Comment by ejgallego** [comment]
Thanks to you @rlepigre , I need to run right now, if you could push your patch to this branch it'd be great.

**Comment by rlepigre** [comment]
> Thanks to you @rlepigre , I need to run right now, if you could push your patch to this branch it'd be great.

I'll try to when I find a minute.

**Comment by Alizter** [comment]
tldr: The `_build` directory is not guaranteed to be at `%{project_root}` but it will be at `%{workspace_root}`. To see this put the `Coq` repo in a directory and put a `dune-workspace` file and run `Dune` from there.

**Comment by rlepigre** [comment]
@ejgallego I pushed our various fixes.

**Comment by SkySkimmer** [comment]
Should be rebased after #19857

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Review by SkySkimmer — APPROVED** [review]
Not sure the coqidetop shim is entirely useful but whatever

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Comment by ejgallego** [comment]
Thanks @rlepigre
